Abstract the DMA and Codec DAI ops for the BE DAI widgets

ranj063 opened the pull request:
Introduce SOF DAI ops to abstract the BE DAI widget DMA, codec_dai and dai_config ops. This change will allow adding the chained DMA mode or the DSP-less mode for the BE DAI widgets with a simple change like this:
```
diff --git a/sound/soc/sof/intel/hda-dai-ops.c b/sound/soc/sof/intel/hda-dai-ops.c
index 1d1e238d032e..1be992367a89 100644
--- a/sound/soc/sof/intel/hda-dai-ops.c
+++ b/sound/soc/sof/intel/hda-dai-ops.c
@@ -433,10 +433,19 @@ void hda_set_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
        }
        case SOF_INTEL_IPC4:
        {
+               struct snd_sof_widget *pipe_widget = swidget->pipe_widget;
                struct sof_ipc4_copier *ipc4_copier = sdai->private;
+               struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+               struct snd_sof_dai_dma_ops *dma_ops = sdai->ops->dma_ops;
+
+               if (ipc4_copier->dai_type != SOF_DAI_INTEL_HDA)
+                       break;
+
+               if (pipeline->use_chain_dma) {
+                       dma_ops->pre_trigger = NULL;
+                       dma_ops->post_trigger = NULL;
+               }
 
-               if (ipc4_copier->dai_type == SOF_DAI_INTEL_HDA)
-                       sdai->ops = &hda_ipc4_dai_ops;
                break;
        }
        default:
```

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -243,7 +239,8 @@ static int hda_link_dma_trigger(struct snd_pcm_substream *substream, int cmd)
 	case SNDRV_PCM_TRIGGER_STOP:
-		ret = hda_link_dma_cleanup(substream, hext_stream, cpu_dai, codec_dai, true);
+		snd_hdac_ext_stream_clear(hext_stream);
+		ret = hda_link_dma_cleanup(substream, hext_stream, cpu_dai, codec_dai);
```
I am pretty sure we added that 'trigger_stop' flag on purpose in the past to keep consistent sequences between different cases. I don't mind undoing this, but this wasn't a random decision.

ranj063 replied:
@plbossart whatever the reason for the split was, this is needed to unify the DMA trigger op for IPC3 and IPC4.

ujfalusi replied:
The commit subject and message is incorrect:
This change is a no change in functionality, you drop the `bool trigger_suspend_stop` parameter from `hda_link_dma_cleanup()` and that's it.

plbossart replied:
> @plbossart whatever the reason for the split was, this is needed to unify the DMA trigger op for IPC3 and IPC4.

Is this a goal?

I thought the main goal was to unify the HDAudio-based DAI handling (existing, DSP less, future extensions). 

Do we want to unify the DMA trigger for IPC3 and IPC4? Or is this made possible by the refactoring?

A bit confused here.

ujfalusi replied:
The change itself looks sane from the change pow, regardless of what the bigger aim is.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -158,10 +158,9 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
 
 static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
-				  struct snd_pcm_hw_params *params)
+				  struct snd_pcm_hw_params *params, struct snd_soc_dai *cpu_dai)
```
this is also a bit controversial, in initial cleanups we made sure the DMA, DAI and IPC stuff were relatively isolated. I am not sold on the idea of dealing with DMA and DAI in the same functions.

ranj063 replied:
but you are dealing with them anyway. All I am saying use the pointer thats already available instead of letting it go and then looking it up again in this function

ujfalusi replied:
iow, you are propagating the `dai` pointer we got from ASoC core down instead of getting a dai via `asoc_rtd_to_cpu(rtd, 0);`, right?

A commit message update might clarify this change?

ujfalusi replied:
I expect that the `dai` is always == `asoc_rtd_to_cpu(rtd, 0)`

plbossart replied:
The use of cpu_dai is only to store the stream tag/hext_stream information.

What really matters in this function is the codec dai which is use to determine what the link configuration will be.

I am not clear on the merit of adding a new parameter when we can follow the same approach for cpu_dai and codec_dai.

Is this to avoid looking up the cpu_dai multiple times?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -475,7 +473,7 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 		pipeline->state = SOF_IPC4_PIPE_RESET;
 
-		ret = hda_link_dma_cleanup(substream, hext_stream, cpu_dai, codec_dai);
+		ret = hda_link_dma_cleanup(substream, hext_stream, dai, codec_dai);
```
why not call it cpu_dai in the parameter list then? 'dai' adds ambiguity on why there's no prefix and there's one for 'codec_dai'.

ranj063 replied:
but why would I change the signature of the function unnecessarily? If you ook at the DAI drv ops, it is called dai and its implicit that it is the CPU DAI already

ujfalusi replied:
Since this driver is for the CPU side, it is the CPU_DAI, yes, that's correct.

plbossart replied:
I really don't see how going from 'cpu_dai, codec_dai' to 'dai, codec_dai' helps make the code more readable. It just confusing for the reviewer to see differences like this.

ranj063 replied:
I have to insist to keep "dai" here even it is confusing for you, Pierre. "dai" is the CPU dai in all of the DAI drv ops callbacks. I see no merit in changing this to cpu_dai

plbossart replied:
the point is that naming matters and reviewers shouldn't have to think on which objects the code changes.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,57 @@ struct snd_sof_route {
+ * @reset: Function pointer for DAI DMA reset
+ * @pre_trigger: Function pointer for DAI DMA pre-trigger actions
+ * @trigger: Function pointer for DAI DMA trigger actions
+ * @post_trigger: Function pointer for DAI DMA post-trigger actions
```
optional? mandatory?

ranj063 replied:
All of our ops in this file are optional unless specified otherwise. I'm not sure we need to add optional to all the ops. It would just make it too verbose.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,57 @@ struct snd_sof_route {
+ * @get_dma_data: Function pointer to get the DAI DMA data
+ * @put_dma_data: Function pointer to release the DAI DMA data
+ * @setup: Function pointer for DAI DMA setup
+ * @reset: Function pointer for DAI DMA reset
```
usually we use setup/free. setup/reset sounds very odd.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,57 @@ struct snd_sof_route {
+/**
+ * struct snd_sof_dai_dma_ops - DAI DMA ops
+ * @get_dma_data: Function pointer to get the DAI DMA data
+ * @put_dma_data: Function pointer to release the DAI DMA data
```
comment is confusing, is this different from DAI dma_data?

ranj063 replied:
but these are common ops. For HDA DAI it is dma_data but should I use the variable name here? DMA data is quite clear isnt it?

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,179 @@
+#include "../sof-audio.h"
+#include "hda.h"
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
```
that define is not going to work for newer platforms.

ranj063 replied:
FOr now this only applies to HDA, so this is perfectly valid right? We will have to deal with the changes when we get to it right?

plbossart replied:
It wouldn't hurt to add a comment here to warn the reviewer of future/known changes.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,179 @@
+}
+
+static void hda_put_dma_data(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			     struct snd_pcm_substream *substream)
```
ok so it's a lot more than just get/put, this is really about managing link DMAs, including tag assignment and release. The naming could be improved IMHO.

maybe stream_get_dma and stream_put_dma?

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -144,8 +144,72 @@ static struct snd_sof_dai_dma_ops hda_dma_ops = {
 
+static int hda_codec_dai_params(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+				void *dma_data, struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params, unsigned int *val)
```
"val" is not a very self-explanatory variable name, is it?

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -139,9 +139,27 @@ static void hda_put_dma_data(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_d
+{
+	struct hdac_ext_stream *hext_stream = dma_data;
+
+	snd_hdac_ext_stream_setup(hext_stream, val);
```
what was wrong with 'format_val'?

ranj063 replied:
@plbossart format_va is very HDA specific. WIth other DAI's especically, non-Intel, it could be anything. So why should we be restricting?

plbossart replied:
this is an implementation for Intel DAIs, and the format is not going away.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -155,6 +155,28 @@ static void hda_dma_reset(struct snd_sof_dev *sdev, void *dma_data)
 }
 
+static int hda_dai_config(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *w,
+			  int channel, unsigned int flags)
```
what does 'channel' represent?

ranj063 replied:
channel represents the link DMA ID for HDA, link ID for SDW etc

ranj063 replied:
this change is removed now. I've reused the existing tplg IPC op instead

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -585,7 +481,7 @@ void hda_set_dai_drv_ops(struct snd_sof_dev *sdev, struct snd_sof_dsp_ops *ops)
 			    strstr(ops->drv[i].name, "Analog") ||
 			    strstr(ops->drv[i].name, "Digital"))
-				ops->drv[i].ops = &ipc4_hda_dai_ops;
+				ops->drv[i].ops = &hda_dai_ops;
```
this is a bit confusion, if the ops are the same between IPC and IPC4 why do wee need the switch()?

ranj063 replied:
for SSP and the NHLT for IPC4

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -319,23 +302,37 @@ static int hda_dai_suspend(struct hdac_bus *bus)
+			if (ops && ops->dma_ops && ops->dma_ops->post_trigger) {
+				ret = ops->dma_ops->post_trigger(sdev, cpu_dai,
+								 hext_stream->link_substream,
+								 SNDRV_PCM_TRIGGER_SUSPEND);
```
I didn't really follow how the hw_free and post-trigger were related. The naming points to different phases in the ALSA world.

ranj063 replied:
For IPC3, post_trigger does what hda_dai_hw_free_ipc(). For IPC4 also, we were doing the same and it made no sense. SO this patch actually fixes it

plbossart reviewed (commented):
@ranj063 I think you're on to something with the .pre and .post trigger.

I added a bunch of comments to try and help clarify the code, most of the changes are confusing due to naming aliases to different concepts.

bardliao commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -168,50 +161,19 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
 	struct hdac_ext_link *hlink;
 	struct snd_sof_dev *sdev;
 	struct hdac_bus *bus;
+	unsigned int format_val, link_bps;
```
one variable per line?

ujfalusi replied:
they are not initialized, so this is fine.

plbossart replied:
it's still better to have one variable per line for consistency.

bardliao commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,111 +29,27 @@ MODULE_PARM_DESC(sof_use_tplg_nhlt, "SOF topology nhlt override");
 	struct hdac_stream *hstream = &hext_stream->hstream;
 	struct hdac_bus *bus = hstream->bus;
 	struct sof_intel_hda_stream *hda_stream;
+
```
extra line

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -225,24 +187,34 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
+	dev_dbg(bus->dev, "format_val=%d, rate=%d, ch=%d, format=%d\n",
+		format_val, params_rate(params), params_channels(params), params_format(params));
+
+	snd_hdac_ext_stream_setup(hext_stream, format_val);
```
the order of snd_hdac_ext_stream_reset(), snd_hdac_ext_stream_setup() and snd_hdac_ext_bus_link_set_stream_id() does not matter?

ranj063 replied:
doesnt  matter as far as I can see. and refactoring it like this makes it easier to call the ops later on

plbossart replied:
it's always hdac_stream_reset then hdac_stream_setup @ranj063 

the set_stream_id() touches the multi-link registers so there's no specific order with the stream parts.

bardliao commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -155,11 +155,123 @@ static void hda_dma_free(struct snd_sof_dev *sdev, void *dma_data)
+
+	w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
+	swidget = w->dobj.private;
+	pipe_widget = swidget->pipe_widget;
```
This will be conflict with the `ASoC: SOF: Introduce struct snd_sof_pipeline` commit

ranj063 replied:
fixed

RanderWang commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -266,6 +266,33 @@ static int hda_ipc4_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
+		unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_FREE;
+
+		flags |= SOF_DAI_CONFIG_FLAGS_NONE << SOF_DAI_CONFIG_FLAGS_QUIRK_SHIFT;
+		data.dai_data = DMA_CHAN_INVALID;
```
do we need to init data.dai_index ?

ranj063 replied:
no dai_index is not used with HDA DAIs

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -187,46 +187,41 @@ static int hda_dai_widget_update(struct snd_soc_dapm_widget *w,
+	if (tplg_ops && tplg_ops->dai_config) {
+		unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
+
+		flags |= SOF_DAI_CONFIG_FLAGS_2_STEP_STOP << SOF_DAI_CONFIG_FLAGS_QUIRK_SHIFT;
```
also here for data.dai_index init ?

ranj063 replied:
dai_index is only needed for SDW and it doesnt use these ops

RanderWang commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,54 @@ struct snd_sof_route {
+	int (*post_trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			    struct snd_pcm_substream *substream, int cmd);
+};
+
```
is it not for sdw dai ?

ranj063 replied:
not right now

plbossart replied:
it's a good point though, the naming can be confusing for SoundWire. Although for SoundWire the pre and post trigger are the only things that can be configured - the trigger is handled with standard registers - and it's at the device level, not the dai level.

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,54 @@ struct snd_sof_route {
+	void *(*stream_get_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				struct snd_pcm_substream *substream);
+	void (*stream_put_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			       struct snd_pcm_substream *substream);
```
The 'stream' part of the callback does not add information on top of the `get`/`put_dma`, but the get/put is a bit hard to put in place. `get_dma_data()` perhaps?

ranj063 replied:
thats was I had get_dma_data/put_dma_data. This was @plbossart 's suggestion. If we have majority/consenses, I can change it whatever way you agree upon

plbossart replied:
@ujfalusi I wanted to have a naming difference with ASoC core functions, e.g. snd_soc_dai_get_dma_data()

This is not a wrapper for those functions, so we have improve both the naming and the description. What does 'get' mean here?

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,413 @@
+			dma_ops = sdai->ops->dma_ops;
+
+			/* set the post DMA trigger op */
+			dma_ops->post_trigger = hda_ipc3_post_trigger;
```
Why post? If you have one of the three trigger ops used it does not really matter which one you pick they are going to be in sequence and if there is only one used then it is the `trigger` itself as post would imply that there is something which precedes this call.

ranj063 replied:
so you'd get here for the case where you have paused a stream but the system suspends. The PAUSE trigger is noth IPC3 and IPC4 would have stopped the DMA. What we wouldnt have done is the post_trigger() is to free the DAI CONFIG for IPC3 and reset the pipeline for IPC4. And post_trigger() handles that

plbossart replied:
comments needed here!

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,413 @@
+
+			/* set the pre/post trigger ops */
+			dma_ops->pre_trigger = hda_ipc4_pre_trigger;
+			dma_ops->post_trigger = hda_ipc4_post_trigger;
```
The pre is the trigger? If there is no trigger then the pre_trigger becomes pre_post_trigger..

ranj063 replied:
trigger is already in the hda_dma_ops struct

ranj063 replied:
maybe i should move it here to not confuse you

ujfalusi replied:
Thanks!

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -473,6 +521,7 @@ struct snd_sof_dai {
 	int number_configs;
 	int current_config;
 	struct list_head list;	/* list in sdev dai list */
+	const struct snd_sof_dai_ops *ops;
```
I would argue that this is Intel only and can be simply managed internally within sound/soc/sof/intel just fine by using an array for the second level of callbacks.

In essence the change would be similar but w/o touching core sof files.

ranj063 replied:
@ujfalusi I would have made this Intel-only too but look at chained-dma mode. It applies to individual pipelines and not the blanket topology like the dsp-less mode. So I need the ops to be different for the chained-dma DAI widget and the regular pipeline-based DAI widget.

ujfalusi replied:
I still believe this is a Intel only thing and there should not be changes in core for this.

If the argument/reason for this is to pave the way for the ChainDMA support, then I still think this is not a correct way.

The ChainDMA for the stream/path/dai is coming from the topology, so using a flag in core components to say that this DAI should be used with ChainDMA is OK.

In Intel code the `hda_dai_get_ops()` can still look up the dai widget, check the flag and decide which set of callback ops it will return.
This series then would not need to touch the core and focus on what's important here: re-arrange the Intel specific HDA code for new features.

I still argue that the core should not know anything about this peculiarity.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,413 @@
+
+	hext_stream = snd_soc_dai_get_dma_data(cpu_dai, substream);
+	snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
+	snd_hdac_ext_stream_release(hext_stream, HDAC_EXT_STREAM_TYPE_LINK);
```
I think the intention is that the `snd_hdac_ext_stream_release()` is used with `snd_hdac_ext_stream_assign()` but we have open coded version of the `snd_hdac_ext_stream_assign()` as `hda_link_stream_assign()` Should we either use the matching assign or open code the release?

For symmetry, that is.

plbossart replied:
it's a good question that @ujfalusi is asking, I don't recall what exactly hda_link_stream_assign() was supposed to fix. we may want to revisit this in a second step though,

ujfalusi replied:
Since this is a code move and it worked fine, I'm OK with this, but we should try to remember and re-visit it.

ujfalusi reviewed (commented):
@ranj063, It is a bit big change to see it through but looks nice refactoring.

As I said in one of the comments I think this whole next level of abstraction should not be visible in core level and can be 'hidden' within Intel code where you don't need to carte about too broad abstraction and opt to use matching terms and avoid passing `void *` around just to re-cast it to the same type as it was before.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -159,6 +151,7 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
 	}
 	snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
 	snd_hdac_ext_stream_release(hext_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	snd_hdac_ext_stream_reset(hext_stream);
```
somehow one of my main comments on this patch was removed or not saved, I think this patch should be split in code moves and changes. I am not able to visualize if this change is correct for example. This is a helper used in 3 different parts of the code, this looks scary when mixed with all kinds of changes.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -465,6 +465,54 @@ struct snd_sof_route {
+/**
+ * struct snd_sof_dai_ops - DAI ops
+ * @dma_ops: Pointer to the DAI DMA ops
+ * @codec_ops: Pointer to the CODEC DAI ops
```
oh this is so confusing.

Is this a DAI ops for codec DAIs only?
If yes, why do we care about DMA for codec dai?
If not, why do we care about the codec DAI ops?

Lost.

ranj063 replied:
Its ops for the codec_dai associated with the cpu_dai. Im all ears if you have a better name for it

plbossart commented on sound/soc/sof/ipc3-topology.c:
```diff
@@ -1587,6 +1587,10 @@ static int sof_ipc3_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 
+	/* set DAI ops */
+	if (sof_ops(sdev)->set_dai_widget_ops)
+		sof_ops(sdev)->set_dai_widget_ops(sdev, swidget);
```
what does 'set the ops' mean?

usually we have set/get, but there's no get?

did you mean 'assign_dai_widget_ops'? or 'select_dai_widget_ops'?

ranj063 replied:
I think set is perfectly valid in this situation but Im not going to argue about a name. So will change to select.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -139,9 +139,27 @@ static void hda_stream_put_dma(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu
 	.stream_get_dma = hda_stream_get_dma,
 	.stream_put_dma = hda_stream_put_dma,
+	.setup = hda_dma_setup,
+	.free = hda_dma_free,
```
commit title should be updated to ASoC: SOF: Intel: hda-dai: Add setup/free DAI DMA ops

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -417,13 +426,23 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
-	return hda_dai_hw_free_ipc(substream->stream, dai);
+	if (tplg_ops && tplg_ops->dai_config) {
+		data.dai_data = DMA_CHAN_INVALID;
+		return tplg_ops->dai_config(sdev, swidget, SOF_DAI_CONFIG_FLAGS_NONE, &data);
```
I am really confused on this one. How is this equivalent between left and right?

ranj063 replied:
i've just literally done whatever dai_dai_hw_free_ipc() does here. This will help me get rid of that function eventually

plbossart reviewed (commented):
more comments below @ranj063 

I am rather confused byu the use of the CONFG_FLAGS, this was initially for SSP only, no?

plbossart commented:
conflict @ranj063 ?

bardliao reviewed (dismissed):
I couldn't see any issue from the PR. So, approved

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -204,8 +205,6 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
 		link_bps = codec_dai->driver->capture.sig_bits;
 
-	snd_hdac_ext_stream_reset(hext_stream);
-
```
I'm not sure how this change makes things balanced, but all other uses of `snd_hdac_ext_stream_reset()` are right before a call to `snd_hdac_ext_stream_setup()` and this is how it has been in SOF.

By the look of it, the reset is needed before setup?

ranj063 replied:
@ujfalusi think about this: When you start stream for the first time ever, you dont need reset, but you need it after stopping. So previously we were doing a reset and start every time but now we do a setup during hw_params() for start and reset during cleanup() for stop. It does the same thing but a bit more logically

ujfalusi replied:
@ranj063, all I'm saying that _all_ existing code does the `snd_hdac_ext_stream_reset()` first followed by `snd_hdac_ext_stream_setup()` and never reset the ext_stream _after_.

But the `snd_hdac_ext_stream_reset()` not only resets the HDA Stream Descriptor registers and FIFO but it will also make sure that the stream is out from reset before continuing.

If you say that it is granted that when we are here the SRTS bit for the stream is guarantied to be 0 (which it should be as 0 is the reset value) then OK, but I just wanted to note this change.

plbossart replied:
I am also not following why we would change the programming sequences. Reset is always followed by setup, this is even documented in the state machine I added in sound/hda/hdac_stream.c

Let's try to keep what works, no?

ranj063 commented:
@ujfalusi your wish granted. I moved the ops to be confined within Intel. But I had to add a void * now to hide it behind

ujfalusi commented on sound/soc/sof/intel/hda.h:
```diff
@@ -913,4 +913,53 @@ extern struct sdw_intel_ops sdw_callback;
+	void *(*stream_get_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				struct snd_pcm_substream *substream);
+	void (*stream_put_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			       struct snd_pcm_substream *substream);
```
We call this what it is now? get_hext_stream and return `struct hdac_ext_stream *` ?

ujfalusi commented on sound/soc/sof/intel/hda.h:
```diff
@@ -913,4 +913,53 @@ extern struct sdw_intel_ops sdw_callback;
+	void (*stream_put_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			       struct snd_pcm_substream *substream);
+	int (*setup)(struct snd_sof_dev *sdev, void *dma_data, unsigned int val);
+	void (*free)(struct snd_sof_dev *sdev, void *dma_data);
```
we can also call the dma_data in it's name: `struct hdac_ext_stream *hext_stream`

ujfalusi replied:
dma_data is always hext_stream, why not call it like that?

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -491,6 +491,8 @@ struct snd_sof_dai {
 	int current_config;
 	struct list_head list;	/* list in sdev dai list */
+	/* This is intended for selecting widget specific ops for the BE DAI widgets */
+	void *private_ops;
```
Can we be more generic and `platform_be_data`, but this will also imply that a given DAI can only use a static parameter?
Since ChainDMA support has been mentioned:
A DAI can be used with ChainDMA or it can not be used with ChainDMA? We can not select 'low latency'/'passthrough' (ChainDMA) PCM and a processed or enhanced (DSP processing) PCM towards the same DAI like we do with normal/DeepBuffer?

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -548,6 +549,10 @@ static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 
+	/* set DAI ops */
+	if (sof_ops(sdev)->select_dai_widget_ops)
+		sof_ops(sdev)->select_dai_widget_ops(sdev, swidget);
```
Can't you do this when the PCM is opened based on the same information in Intel code? Or when first time you need this information in Intel code and it is not set yet?

plbossart replied:
Or when the widget is created, why do this in the setup phase where we allocate stuff? This is indeed more of a one-time assignment. Doing this assignment multiple times when we setup is a clear sign this doesn't belong here.

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -258,6 +370,11 @@ void hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *
+
+			dma_ops->pre_trigger = hda_ipc4_pre_trigger;
+			dma_ops->trigger = hda_trigger;
+			dma_ops->post_trigger = hda_ipc4_post_trigger;
```
@ranj063, this is definitely not going to work as intended.
You can not have separate callbacks defined for different DAIs/uses as you are modifying a single `hda_dai_ops` and it's single `hda_dma_ops` if a single DAI would need different ops, you will change it for all DAIs.

You need all this in static or do a kmemdump and modify the copy.

plbossart replied:
not following @ujfalusi's comment

ujfalusi reviewed (commented):
@ranj063, thanks for the changes, but I don't think this is going to work for the purpose you are citing.

I don's see why we need to have the `select_dai_widget_ops()` call from core. It can be done just fine in the code when you are fetching the `struct hda_dai_widget_ops` for the given DAI widget. If it is NULL atm, you go out and use the same logic as you are using to set it (initialize) and that's that.
When you don't need these callbacks I would set an empty `struct hda_dai_widget_ops nop_ops` so that you will skip the lookup again.

It should work in a same way, but no need to have new callback in core where it is not needed.

Also: the modifying of a single ops struct, which is used for _all_ DAIs is not going to work, I would rather define the structs as `static const` for each use and just assign them to the DAIs where they are needed.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -209,9 +176,12 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
 	struct snd_sof_dev *sdev;
 	struct hdac_bus *bus;
+	unsigned int format_val;
+	unsigned char stream_tag;
```
it's not clear why we used 'unsigned char', we seem to use 'int stream_tag' everywhere else?

ranj063 replied:
this is just a code move @plbossart so I kept it as is. But eventually by the end of the ops addition, all stream_tag will be unified to use int

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -151,6 +151,7 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
 	}
 	snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
 	snd_hdac_ext_stream_release(hext_stream, HDAC_EXT_STREAM_TYPE_LINK);
+	snd_hdac_ext_stream_reset(hext_stream);
```
this is not what is documented and is not aligned with all other uses of snd_hdac libraries.

This is not the place to be creative IMHO. This is scary enough as is. Reset is followed by setup. Let's stick to that, shall we?

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -913,4 +913,53 @@ extern struct sdw_intel_ops sdw_callback;
+	int (*trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+		       struct snd_pcm_substream *substream, int cmd);
+	int (*post_trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			    struct snd_pcm_substream *substream, int cmd);
```
this really needs a lot more information on what stream_get_dma does, what setup/free allocate and the purpose of the trigger functions.

Abstracting callbacks should come with explanations/examples on when these callbacks should be used and what for.

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -913,4 +913,53 @@ extern struct sdw_intel_ops sdw_callback;
+ * struct hda_dai_widget_codec_dai_ops - Codec DAI ops optional by default unless specified
+ * otherwise
+ * @params: Function pointer for codec DAI params
+ * @free: Function pointer for codec DAI free
```
is this related to hw_params and hw_free?

plbossart replied:
the comment still stands, what does this accomplish? This is a bit confusing since the codec dai has its own params/free.

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -913,4 +913,53 @@ extern struct sdw_intel_ops sdw_callback;
+ */
+struct hda_dai_widget_ops {
+	struct hda_dai_widget_dma_ops *dma_ops;
+	struct hda_dai_widget_codec_dai_ops *codec_ops;
```
so I am completely confused no why we need to abstract codec_ops. the codec dai is handled by the codec driver.
To use your own comment in the previous patch '"dai" is the CPU dai in all of the DAI drv ops callbacks. I see no merit in changing this to cpu_dai', so why do we care about codec_dais?

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -548,6 +549,10 @@ static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 	if (ret)
 		goto free_copier_config;
 
+	/* set DAI ops */
```
unrelated but like 531 our use of default: is completely wrong. this should be for HDA, and default should be an error case.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -174,14 +92,11 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
-		snd_soc_dai_set_dma_data(cpu_dai, substream, (void *)hext_stream);
-	}
+	if (!hext_stream)
+		return -EBUSY;
```
Humm, this changes things quite a bit. The get_dma() should only be done if there's no DMA already.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -224,11 +139,16 @@ static int hda_link_dma_prepare(struct snd_pcm_substream *substream, struct snd_
 	int ret;
 
+	if (ops && ops->dma_ops && ops->dma_ops->stream_get_dma)
+		hext_stream = ops->dma_ops->stream_get_dma(sdev, cpu_dai, substream);
```
the naming is also a bit weird, get/put are not balanced in this scenario. We only allocate if we don't have a DMA, and use the one we have otherwise. The put releases the DMA regardless of how many times we did a get.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -145,8 +145,72 @@ static struct hda_dai_widget_dma_ops hda_dma_ops = {
+	dev_dbg(bus->dev, "format_val=%d, rate=%d, ch=%d, format=%d\n",
+		*format_val, params_rate(params), params_channels(params), params_format(params));
+
+	return 0;
```
ok, so I accept that this has to be done by using the codec dai information....

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -145,8 +145,72 @@ static struct hda_dai_widget_dma_ops hda_dma_ops = {
+		snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag);
+	}
+
+	return 0;
```
....but here this is completely unrelated to codec dais. The problem is to find the hlink, and we could use the CPU dai name for this. There's no reason why we should use the codec name, this clearly doesn't translate well for other platforms.

plbossart replied:
the comment still stands.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -140,9 +140,27 @@ static void hda_stream_put_dma(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu
+
+	snd_hdac_ext_stream_reset(hext_stream);
+}
+
```
again I think we want to keep reset and setup in this order. Let's not try to think to hard on elegance and symmetry, this isn't how the code is used today.

plbossart reviewed (commented):
I revisited this @ranj063 and there are lots of good things. 

Overall I would err on the side of keeping the known reset/setup sequence as is for hda streams. We don't know what we don't know here, and it's not true that stream reset is the opposite of stream_setup - it's stream cleanup

````
*
 *			   |
 *	stream_init()	   |
 *			   v
 *			+--+-------+
 *			|  unused  |
 *			+--+----+--+
 *			   |    ^
 *	stream_assign()	   | 	|    stream_release()
 *			   v	|
 *			+--+----+--+
 *			|  opened  |
 *			+--+----+--+
 *			   |    ^
 *	stream_reset()	   |    |
 *	stream_setup()	   |	|    stream_cleanup()
 *			   v	|
 *			+--+----+--+
 *			| prepared |
 *			+--+----+--+
 *			   |    ^
 *	stream_start()	   | 	|    stream_stop()
 *			   v	|
 *			+--+----+--+
 *			|  running |
 *			+----------+
 
````

I would also revisit the codec dai dependencies, this is very very specific to HDAudio at the moment.

ranj063 commented:
Im going to wait for @plbossart 's update to calculate the format_val before updating this PR

jsarha commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -0,0 +1,186 @@
+		struct sof_dai_private_data *private = sdai->private;
+
+		if (private->dai_config->type == SOF_DAI_INTEL_HDA) {
+			struct hda_dai_widget_ops *ops = &hda_dai_ops;
```
What is the point of using local ops, instead of "sdai->private_ops = &hda_dai_ops;" ?

jsarha commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -187,46 +187,41 @@ static int hda_dai_widget_update(struct snd_soc_dapm_widget *w,
 
 	if (ops && ops->dma_ops && ops->dma_ops->stream_get_dma)
 		hext_stream = ops->dma_ops->stream_get_dma(sdev, dai, substream);
+
```
Adding this new line should probably be in  "ASoC: SOF: Intel: hda-dai: Add DAI widget ops for HDA DAIs".

jsarha commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -268,7 +266,18 @@ static int hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_soc_d
+		unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
+
+		flags |= SOF_DAI_CONFIG_FLAGS_2_STEP_STOP << SOF_DAI_CONFIG_FLAGS_QUIRK_SHIFT;
+		data.dai_data = hdac_stream(hext_stream)->stream_tag - 1;
```
I am not really sure what plg_ops->dai_config() bellow does with &data, but it looks scary to initialize data.dai_data, but leave data.dai_index uninitialized. Should you add initialization to struct "snd_sof_dai_config_data data" declaration?

ranj063 replied:
@jsarha dai_index is used only for SDW and dai_data is used for HDA. I cant initialize it just for the sake of it. Please check the code in ipc4-topology.c to see if it is used at all or not

jsarha replied:
I think it would still make sense to initialize dai_index to some defined invalid value, to make it clear its not to be used. Also, if both values are never used at the same time, then shouldn't this be a union? But for my part, feel free to resolve the conversation (and the one bellow).

jsarha commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -417,13 +426,23 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+	const struct sof_ipc_tplg_ops *tplg_ops = sdev->ipc->ops->tplg;
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_sof_dai_config_data data;
```
Same here, should this be initialized?

jsarha commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -359,65 +357,35 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
-		ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
-						  SOF_IPC4_PIPE_RESET);
+	if (ops && ops->dma_ops && ops->dma_ops->post_trigger) {
+		ret = ops->dma_ops->post_trigger(sdev, dai, substream, cmd);
```
Ok, so here at least dma_ops pre_trigger(), trigger(), and post_trigger() are called in sequence, without anything happening in between. I trust this not the situation everywhere. It would be really nice to have a big picture about in which order and where these different ops are called in different situations, or at least some examples about the sequence.

jsarha reviewed (commented):
Not really getting the big picture yet, but couple of nit pics instead.

lgirdwood commented:
@plbossart @ranj063 should chain DMA go in before this PR now given the ongoing discussion ?

ranj063 commented:
@plbossart @ujfalusi @bardliao @RanderWang @jsarha Please have a look again. The first 5 patches are from #4012. I will have to rebase after that and #3975 are merged

ujfalusi commented on sound/soc/sof/intel/hda.h:
```diff
@@ -927,4 +927,60 @@ extern struct sdw_intel_ops sdw_callback;
+						      struct snd_pcm_substream *substream);
+	void (*release_hext_stream)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				    struct snd_pcm_substream *substream);
+	int (*setup)(struct snd_sof_dev *sdev, void *dma_data, unsigned int format_val);
```
the setup is just a wrapper to `snd_hdac_ext_stream_setup()` so it is really a `setup_ext_stream()`, it has specific functionality and applicable only with HDA.
Similar stands for free.

If I take a a look from the distance, these callbacks are mandatory _if_ we have HDA and the whole widget_dma_ops is the optional. it should be set for HDA and not needed for SSP/DMIC?

ujfalusi commented on sound/soc/sof/sof-audio.h:
```diff
@@ -493,6 +493,8 @@ struct snd_sof_dai {
 	int current_config;
 	struct list_head list;	/* list in sdev dai list */
+	/* This is intended for selecting widget specific ops for the BE DAI widgets */
+	const void *private_ops;
```
Why do you want to lock it down with a name and comment? Can a `const void *platform_private` just work fine with a comment that core is not touching it and it is up to the platform if it is uses it and how?

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -159,29 +81,32 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
+			hext_stream = ops->dma_ops->assign_hext_stream(sdev, cpu_dai, substream);
 	}
 
+	if (!hext_stream)
```
you _always_ need to have the `get_hext_stream` and `assign_hext_stream` set if hda_link_dma_hw_params() is called.
Why not check it first (in init or once) and fail early?

We are just replacing the direct calls with 1:1 abstraction.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -225,11 +150,16 @@ static int hda_link_dma_prepare(struct snd_pcm_substream *substream, struct snd_
 
+	if (ops && ops->dma_ops && ops->dma_ops->get_hext_stream)
+		hext_stream = ops->dma_ops->get_hext_stream(sdev, cpu_dai, substream);
+
```
Again, the `get_hext_stream` is mandatory. hda_link_dma_trigger() will fail if it is not set.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -258,11 +188,14 @@ static int hda_link_dma_trigger(struct snd_pcm_substream *substream, struct snd_
 
-	hext_stream = snd_soc_dai_get_dma_data(cpu_dai, substream);
+	if (ops && ops->dma_ops && ops->dma_ops->get_hext_stream)
+		hext_stream = ops->dma_ops->get_hext_stream(sdev, cpu_dai, substream);
```
yep, `get_hext_stream` is a must.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,108 +29,28 @@ MODULE_PARM_DESC(sof_use_tplg_nhlt, "SOF topology nhlt override");
-		snd_hdac_ext_stream_decouple_locked(bus, res, true);
+	/* select and set the DAI widget ops if not set already */
+	if (!sdai->private_ops)
+		hda_select_dai_widget_ops(sdev, swidget);
```
In think it would be more intuitive:
```c
const struct hda_dai_widget_ops *
hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_dai *sdai) {}

or

const struct hda_dai_widget_ops *
hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget) {}
...
sdai->private_ops = hda_select_dai_widget_ops(sdev, sdai);
```

ujfalusi commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -150,8 +150,68 @@ static const struct hda_dai_widget_dma_ops hda_dma_ops = {
+		return -EINVAL;
+
+	if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_PLAYBACK) {
+		int stream_tag;
```
The stream_tag tends to be unsigned char, int, u32...
The initialization might fit in the same line?

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -213,46 +213,41 @@ static int hda_dai_widget_update(struct snd_soc_dapm_widget *w,
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
 	const struct hda_dai_widget_ops *ops = hda_dai_get_ops(substream, dai);
+	const struct sof_ipc_tplg_ops *tplg_ops = sdev->ipc->ops->tplg;
```
We have a helper for getting the IPC ops: `sof_ipc_get_ops()`, it might be beneficial to use it?

ujfalusi reviewed (commented):
@ranj063, it looks much better now that the change is contained under the Intel directory, thank you!

I had not realized that we have so big amount of IPC code outside of the core, it is hard to claim that the IPC of SOF is platform agnostic ;)

I have left couple of comments for the changes.

plbossart commented on sound/soc/sof/intel/hda.h:
```diff
@@ -927,4 +927,61 @@ extern struct sdw_intel_ops sdw_callback;
+ * otherwise
+ * @params: Function pointer for codec DAI params
+ * @free: Function pointer for codec DAI free
+ * @get_max_bps: Mandatory function pointer to get the max bps supported by the codec
```
humm, no this should be optional. This is specific to HDaudio, this doesn' exist for other interfaces even if they rely on the HDaudio DMA.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -150,8 +150,66 @@ static const struct hda_dai_widget_dma_ops hda_dma_ops = {
+		int stream_tag = hdac_stream(hext_stream)->stream_tag;
+
+		snd_hdac_ext_bus_link_set_stream_id(hlink, stream_tag);
+	}
```
this is a bit awkward. All we need here is something that return the hlink. The rest to set the stream_id is going to be completely codec-agnostic.

In this case we use the codec_dai only to get the hlink information. We could do this differently by checking the name of the CPU DAI.

plbossart reviewed (changes requested):
This is much better @ranj063 but I think the codec_dai ops are probably overkill. The only piece of information we need is maxbps, and only in the case of HDaudio DAIs. the rest that's needed is just a mechanism to get the hlink, set the stream and clear the stream.

ranj063 commented:
> This is much better @ranj063 but I think the codec_dai ops are probably overkill. The only piece of information we need is maxbps, and only in the case of HDaudio DAIs. the rest that's needed is just a mechanism to get the hlink, set the stream and clear the stream.

@plbossart I've dropped the codec_ops for now. I think we should revisit this later

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -231,33 +231,17 @@ static int hda_dai_widget_update(struct snd_soc_dapm_widget *w,
-	/* set up the DAI widget and send the DAI_CONFIG with the new tag */
-	return hda_dai_widget_update(w, stream_tag - 1, true);
-}
-
```
I am not following here, we had a helper that was invoked from two different places, and now we end-up with duplicated code in two places. what's the gain?

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -277,12 +277,41 @@ static const struct hda_dai_widget_dma_ops hda_ipc4_dma_ops = {
+		flags |= SOF_DAI_CONFIG_FLAGS_NONE << SOF_DAI_CONFIG_FLAGS_QUIRK_SHIFT;
+		data.dai_data = DMA_CHAN_INVALID;
+		return tplg_ops->dai_config(sdev, swidget, flags, &data);
+	}
```
and what looks like a duplication here as well for dai_config?

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -124,30 +60,87 @@ static int sdw_params_stream(struct device *dev,
 
-	return hda_ctrl_dai_widget_setup(w, SOF_DAI_CONFIG_FLAGS_NONE, &data);
+	if (tplg_ops && tplg_ops->dai_config) {
+		unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
```
not following, on the left we have FLAGS_NONE and here HW_PARAMS?

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -124,30 +60,87 @@ static int sdw_params_stream(struct device *dev,
+		int ret;
+
+		/* set HW_FREE flag along with any quirks */
+		flags = SOF_DAI_CONFIG_FLAGS_HW_FREE;
```
same here, why change the flags?

plbossart reviewed (commented):
The abstraction for the dai widget looks good @ranj063 but I find the code related to dai_config a bit messy. It looks like code duplication is increased, and the flags seem changed at times?

This should be relatively easy to clean-up, you've done the hard part already.

ranj063 commented:
> The abstraction for the dai widget looks good @ranj063 but I find the code related to dai_config a bit messy. It looks like code duplication is increased, and the flags seem changed at times?
> 
> This should be relatively easy to clean-up, you've done the hard part already.

@plbossart Basically I want to remove the functions hda_ctrl_widget_setp/free(). They were left over from a previous implementation where we actually setup/freed the widgets. But I agree I introduced too much duplication and now I've fixed those. I think the new version makes it a lot more intuitive to see what flags are sent where.

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -277,12 +277,37 @@ static const struct hda_dai_widget_dma_ops hda_ipc4_dma_ops = {
 };
 
+static int hda_ipc3_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				 struct snd_pcm_substream *substream, int cmd)
```
the commit message describes two independent parts, could the IPC4 fix for suspend while paused be handled with a dedicated commit?

ranj063 replied:
Unfortunately no I cant split that part. The use of post_trigger() in suspend is an automatic fix for IPC4 that we get for free.

plbossart replied:
ok

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -159,29 +92,31 @@ static int hda_link_dma_cleanup(struct snd_pcm_substream *substream,
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_soc_dai *codec_dai = asoc_rtd_to_codec(rtd, 0);
-	struct hdac_ext_stream *hext_stream;
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -258,11 +197,18 @@ static int hda_link_dma_trigger(struct snd_pcm_substream *substream, struct snd_
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_soc_dai *codec_dai = asoc_rtd_to_codec(rtd, 0);
-	struct hdac_ext_stream *hext_stream;
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -287,18 +233,18 @@ static int hda_dai_hw_params_update(struct snd_pcm_substream *substream,
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -307,10 +253,17 @@ static int hda_dai_hw_params(struct snd_pcm_substream *substream,
-				snd_soc_dai_get_dma_data(dai, substream);
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -342,13 +295,14 @@ static int hda_dai_config_pause_push_ipc(struct snd_soc_dapm_widget *w)
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -417,27 +371,27 @@ static int ipc3_hda_dai_trigger(struct snd_pcm_substream *substream,
-	struct sof_ipc4_pipeline *pipeline;
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct hdac_ext_stream *hext_stream = NULL;
```
useless init?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,6 +29,26 @@ MODULE_PARM_DESC(sof_use_tplg_nhlt, "SOF topology nhlt override");
+	if (tplg_ops && tplg_ops->dai_config) {
+		ret = tplg_ops->dai_config(sdev, swidget, flags, data);
+		if (ret < 0) {
+			dev_err(sdev->dev, "DAI config failed for widget %s\n", w->name);
```
maybe add the flags as well, since we can have multiple cases of dai_config for the same widget?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -273,35 +276,22 @@ static int hda_dai_hw_params(struct snd_pcm_substream *substream,
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
-	const struct sof_ipc_tplg_ops *tplg_ops = sdev->ipc->ops->tplg;
-	int ret = 0;
+	hext_stream = ops->get_hext_stream(sdev, dai, substream);
```
useless init on line 261?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -314,7 +304,12 @@ static int hda_dai_prepare(struct snd_pcm_substream *substream, struct snd_soc_d
 		return ret;
 
-	return hda_dai_hw_params_update(substream, &rtd->dpcm[stream].hw_params, dai);
+	hext_stream = ops->get_hext_stream(sdev, dai, substream);
```
your've already fetched hext_stream line 297

also useless init line 292?

plbossart commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -155,7 +155,129 @@ static void hda_reset_hext_stream(struct snd_sof_dev *sdev, struct hdac_ext_stre
+static int hda_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+		       struct snd_pcm_substream *substream, int cmd)
+{
+	struct hdac_ext_stream *hext_stream = snd_soc_dai_get_dma_data(cpu_dai, substream);
```
get_hext_stream?

Wondering if all use of dma_data should be hidden now in the helpers?

ranj063 replied:
this is for IPC3 at this point which doesnt use the trigger ops and will go away after I unify the flows for IPC3 and IPC4 in the following patches.

plbossart reviewed (commented):
much better indeed @ranj063, I like the updated hda_dai_config, it makes the code a lot less cryptic.

A couple of suggested improvements below, mostly nit-picks.

This is starting to look really good. Maybe we can enjoy the Turkey break then.

plbossart commented:
Not sure what happens with https://sof-ci.01.org/linuxpr/PR3972/build1973/devicetest/index.html?model=TGLU_UP_HDA_ZEPHYR&testcase=check-suspend-resume-5, this seems to be the only fail on HDaudio platforms?

Worth investigating or redoing the tests.

plbossart reviewed (dismissed):
LGTM, we should probably do more testing overnight to see if we have any regressions, this is a rather intrusive patch.

Nice work @ranj063, I think this will help a lot clarify the sequences which were diluted with different ops and branches left and right.

plbossart commented:
@ujfalusi @RanderWang @bardliao @kv2019i @jsarha please prioritize this PR if you have a bit of time on Thursday 16, we are trying to complete this work before the Thanksgiving break. Friday will be too late for any interaction.

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,6 +29,27 @@ MODULE_PARM_DESC(sof_use_tplg_nhlt, "SOF topology nhlt override");
+	if (tplg_ops && tplg_ops->dai_config) {
+		ret = tplg_ops->dai_config(sdev, swidget, flags, data);
+		if (ret < 0) {
+			dev_err(sdev->dev, "DAI config with flags %x failed for widget %s\n",
```
%#x to print hex number ? It does not block me.

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -273,35 +277,22 @@ static int hda_dai_hw_params(struct snd_pcm_substream *substream,
-	int stream = substream->stream;
+	struct snd_sof_dai_config_data data = { 0 };
+	unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
 	int ret;
```
I like the layout
````
unsigned int flags = SOF_DAI_CONFIG_FLAGS_HW_PARAMS;
struct snd_sof_dai_config_data data = { 0 };
struct hdac_ext_stream *hext_stream;
````

RanderWang commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -462,73 +360,62 @@ static int hda_dai_suspend(struct hdac_bus *bus)
-	}
-	default:
-		break;
+		ipc4_data->nhlt = intel_nhlt_init(sdev->dev);
```
why we need to init nhlt in hda_set_dai_drv_ops ?  It is not a issue in this PR.  we remove the switch for ipc version but leave one if for ipc version

RanderWang reviewed (approved):
LGTM

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -29,108 +29,39 @@ MODULE_PARM_DESC(sof_use_tplg_nhlt, "SOF topology nhlt override");
-		snd_hdac_ext_stream_decouple_locked(bus, res, true);
+		/* check if mandatory ops are set */
+		if (!ops || !ops->get_hext_stream)
+			return NULL;
```
nitpick: I would print error when the `get_ext_stream` is missing to make it more obvious.

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -168,20 +102,21 @@ static int hda_link_dma_hw_params(struct snd_pcm_substream *substream,
+	if (!hext_stream) {
+		if (ops->assign_hext_stream)
+			hext_stream = ops->assign_hext_stream(sdev, cpu_dai, substream);
 	}
```
You got me here...
Only `get_hext_stream` is mandatory, which is to to retrieve the dma_data set by the optional `assign_hext_stream`.
If the `assign_hext_stream` is not set (optional!) and this is the first call here then the `get_hext_stream` is going to return with NULL and you are not going to get anything but NULL...

What I was saying is that most, if not all callbacks in `struct hda_dai_widget_dma_ops` are in fact mandatory, don't they?

ujfalusi replied:
also to note: you also must have the `release_hext_stream` defined if you have the `assign_hext_stream`, but you must need to have the `assign_hext_stream` in order to have working stack.

ujfalusi replied:
or you should not return with error in line 118?

ranj063 replied:
The conclusion for this thread is that assign/release need to be optional because couple mode will not need them

ujfalusi commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -417,27 +371,27 @@ static int ipc3_hda_dai_trigger(struct snd_pcm_substream *substream,
-	struct sof_ipc4_pipeline *pipeline;
+	const struct hda_dai_widget_dma_ops *ops = hda_dai_get_ops(substream, dai);
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct hdac_ext_stream *hext_stream;
```
nitpick: would look better two lines down..

ujfalusi reviewed (commented):
@ranj063, this looks really nice!
The only thing holding me to Approve is that I can not follow optionality in `struct hda_dai_widget_dma_ops`.
`assign_hext_stream` is as mandatory as the `get_hext_stream`, which also brings the `release_hext_stream` as mandatory.
By the look, the other ops are really optional.

ujfalusi reviewed (approved):
@ranj063, I'm OK with few fixups (checkpatch).
Let's see how this will behave in the wild and under the DSPless and ChainDMA cases.
